lib/base_service.ts

```diff
@@ -129,12 +145,58 @@ export class BaseService {
     if (this._options.api_key) {
       _credentials.api_key = this._options.api_key;
     }
+    if (this._options.platform_api_key) {
+      _credentials.platform_api_key = this._options.platform_api_key;
+    }
     if (this._options.url) {
       _credentials.url = this._options.url;
     }
     return _credentials;
   }
 
+/**
+ * Wrapper around `sendRequest` that manages access tokens if applicable
+ * 1. If tokens are unsupported or not being used, call `sendRequest`
+ * 2. If tokens are being used but no token is saved, request an access token
+ * 3. If a token is saved but it is expired, refresh the access token
+ * 4. If token is saved and valid, use it to authenticate the service request
+ *
+ * @param {Object} parameters - service request options passed in by user
+ * @param {Function} _callback - callback function to pass the reponse back to
+ * @returns {ReadableStream|undefined}
+ */
+  protected createRequest(parameters, _callback) {
+    // condition 1: using basic or other auth, no need for any tokens
+    if (this.notUsingTokens()) {
+      return sendRequest(parameters, _callback);
+
+      // condition 2: no token has been retrieved, make a call to get the token
+    } else if (this.needToRequestToken()) {
+      tokenRequest(this._options.platform_api_key, (tokenResponse) => {
+        this.setTokenObject(tokenResponse);
+        parameters.defaultOptions.headers.Authorization =
+          `Bearer ${this.tokenInfo.access_token}`;
+        return sendRequest(parameters, _callback);
+      });
+
+      // condition 3: the token was retrieved, but has expired
+    } else if (this.tokenIsExpired()) {
+      tokenRequest(this.tokenInfo.refresh_token, (tokenResponse) => {
+        this.setTokenObject(tokenResponse);
+        parameters.defaultOptions.headers.Authorization =
+          `Bearer ${this.tokenInfo.access_token}`;
+        return sendRequest(parameters, _callback);
+        // this 'true' argument specifies that we're refreshing an existing token
+      }, true);
+
+      // condition 4: the token is good to go, use it as is
+    } else {
+      parameters.defaultOptions.headers.Authorization =
+        `Bearer ${this.tokenInfo.access_token}`;
+      return sendRequest(parameters, _callback);
+    }
+  }
```

Comment:
This code block seems pretty specific to IAM tokens. I wonder if there's an opportunity to create a more general credentials manager? Then you could have a file like `IAMTokenManager` that contains the logic for managing IAM tokens, rather than the base service class.

All of this code is internal, though, so that could always be abstracted later.

Reply:
This is a good idea and something I thought about when I first started working on it. At first thought, it seems like it may be more effort than it's worth as it doesn't seem likely that we'll need to support multiple token types in the near future.

I'm going to push the new code I have to be reviewed for now, but I will spend some time thinking about how we might implement this more generally.